c10/xpu/impl/XPUGuardImpl.h

```diff
@@ -0,0 +1,131 @@
+  }
+
+  DeviceType type() const override {
+    return DeviceType::XPU;
+  }
+
+  Device exchangeDevice(Device d) const override {
+    TORCH_INTERNAL_ASSERT(d.is_xpu());
+    auto old_device_index = c10::xpu::exchange_device(d.index());
+    return Device(DeviceType::XPU, old_device_index);
+  }
+
+  Device getDevice() const override {
+    auto device = c10::xpu::current_device();
+    return Device(DeviceType::XPU, device);
+  }
+
+  c10::optional<Device> uncheckedGetDevice() const noexcept {
+    auto device = c10::xpu::current_device();
+    return Device(DeviceType::XPU, device);
+  }
+
+  void setDevice(Device d) const override {
+    TORCH_INTERNAL_ASSERT(d.is_xpu());
+    c10::xpu::set_device(d.index());
+  }
+
+  void uncheckedSetDevice(Device d) const noexcept override {
+    c10::xpu::set_device(d.index());
+  }
+
+  Stream getStream(Device d) const noexcept override {
+    return getCurrentXPUStream(d.index()).unwrap();
+  }
+
+  Stream getStreamFromGlobalPool(Device d, bool isHighPriority = false)
+      const override {
+    return getStreamFromPool(isHighPriority, d.index());
+  }
+
+  // NB: These do NOT set the current device
+  Stream exchangeStream(Stream s) const noexcept override {
+    XPUStream stream(s);
+    auto old_stream = getCurrentXPUStream(s.device().index());
+    setCurrentXPUStream(stream);
+    return old_stream.unwrap();
+  }
+
+  DeviceIndex deviceCount() const noexcept override {
+    return c10::xpu::device_count();
+  }
+
+  // Event-related functions
+  void destroyEvent(void* event, const DeviceIndex device_index)
+      const noexcept override {}
+
+  void record(
+      void** event,
+      const Stream& stream,
+      const DeviceIndex device_index,
+      const EventFlag flag) const override {
+    TORCH_CHECK(
+        device_index == -1 || device_index == stream.device_index(),
+        "Event device index ",
+        device_index,
+        " does not match recording stream's device index ",
+        stream.device_index(),
+        ".");
+
+    auto* xpu_event = reinterpret_cast<sycl::event*>(*event);
+    XPUStream xpu_stream{stream};
+    *xpu_event = xpu_stream.queue().ext_oneapi_submit_barrier();
+  }
+
+  void block(void* event, const Stream& stream) const override {
+    if (!event)
+      return;
+    auto* xpu_event = reinterpret_cast<sycl::event*>(event);
+    std::vector<sycl::event> event_list{*xpu_event};
+    XPUStream xpu_stream(stream);
+    xpu_stream.queue().ext_oneapi_submit_barrier(event_list);
+  }
+
+  bool queryEvent(void* event) const override {
+    using namespace sycl::info;
+    if (!event)
+      return true;
+    auto* xpu_event = reinterpret_cast<sycl::event*>(event);
+    return xpu_event->get_info<event::command_execution_status>() ==
+        event_command_status::complete;
+  }
+
+  // Stream-related functions
+  bool queryStream(const Stream& stream) const override {
+    XPUStream xpu_stream{stream};
+    // return xpu_stream.query();
+    return false;
```

Comment:
Add comments to explain why always false here.

Reply:
Change to `return xpu_stream.query()`, will raise a `RuntimeError` currently.